hexrdgui/masking/mask_manager.py

```diff
@@ -0,0 +1,390 @@
+)
+from hexrdgui.masking.create_raw_mask import (
+    recompute_raw_threshold_mask, create_raw_mask, rebuild_raw_masks
+)
+from hexrdgui.hexrd_config import HexrdConfig
+from hexrdgui.masking.mask_compatability import load_masks_v1_to_v2
+from hexrdgui.singletons import QSingleton
+from hexrdgui.utils import unique_name
+
+from hexrd.instrument import unwrap_dict_to_h5
+
+from abc import ABC, abstractmethod
+
+
+class Mask(ABC):
+    def __init__(self, name='', mtype='', visible=True):
+        self.type = mtype
+        self.name = name
+        self.visible = visible
+        self.masked_arrays = None
+
+    def update_mask_visibility(self, visibility):
+        self.visible = visibility
+
+    # Abstract methods
+    @abstractmethod
+    def get_data(self):
+        pass
+
+    @abstractmethod
+    def set_data(self, data):
+        pass
+
+    @abstractmethod
+    def update_masked_arrays(self):
+        pass
+
+    @abstractmethod
+    def serialize(self):
+        pass
+
+    @classmethod
+    def deserialize(cls, data):
+        return cls(data['name'], data['mtype'], data['visible'])
+
+
+class RegionMask(Mask):
+    def __init__(self, name='', mtype='', visible=True):
+        super().__init__(name, mtype, visible)
+        self._raw = None
+
+    def get_data(self):
+        return self._raw
+
+    def set_data(self, data):
+        self._raw = data
+        self.update_masked_arrays()
+
+    def update_masked_arrays(self, view=ViewType.raw):
+        if view == ViewType.raw:
+            self.masked_arrays = create_raw_mask(self._raw)
+        else:
+            self.masked_arrays = create_polar_mask_from_raw(self._raw)
+
+    def serialize(self):
+        data = {
+            'name': self.name,
+            'mtype': self.type,
+            'visible': self.visible,
+        }
+        for i, (det, values) in enumerate(self._raw):
+            data.setdefault(det, {})[str(i)] = values
+        return data
+
+    @classmethod
+    def deserialize(cls, data):
+        new_cls = cls(data['name'], data['mtype'], data['visible'])
+        raw_data = []
+        for det in HexrdConfig().detector_names:
+            if det not in data.keys():
+                continue
+            raw_data.extend([(det, v) for v in data[det].values()])
+        new_cls.set_data(raw_data)
+        return new_cls
+
+
+class ThresholdMask(Mask):
+    def __init__(self, name='', mtype='', visible=True):
+        super().__init__(name, mtype, visible)
+        self.min_val = -math.inf
+        self.max_val = math.inf
+        self._hidden_mask_data = None
+
+    def update_mask_visibility(self, visible):
+        if visible == self.visible:
+            return
+
+        self.visible = visible
+        if visible and self._hidden_mask_data:
+            self.set_data(self._hidden_mask_data)
+            self._hidden_mask_data = None
+        elif not visible:
+            self._hidden_mask_data = self.get_data()
+            self.set_data([-math.inf, math.inf])
```

Comment:
I think we should always keep `self.min_val` and `self.max_val` to be whatever they were set as, regardless of the visibility. Elsewhere, functions like `recompute_raw_threshold_mask()` can just check if the mask is visible, and not apply it if it is not visible.